### ISSUE TITLE
Hibernate 6 upgrade: unable to extract query parameter name in QueryCacheKey

### ISSUE THREAD

**Issue by sguilhen** [keycloak/keycloak#16332]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

storage

### Describe the bug

After upgrading to hibernate 6, the `QueryCacheKey` constructor fails due to the query parameter's name being now null. Switching to position doesn't help as that also returns a null value.

This code was put in place in the quarkus3 branch to be able to start Keycloak with the map store: (`JpaMapKeycloakTransaction.QueryCacheKey`)

```
            for (Parameter<?> parameter : query.getParameters()) {
                // HACK TO GET KC TO START WITH NEW STORE - HAS TO BE FIXED AS SOME SEARCHES USE A DIFFERENT EXPRESSIBLE IMPL
                JpaCriteriaParameter criteriaParameter = (JpaCriteriaParameter) parameter;
                SqmBasicValuedSimplePath expressible = (SqmBasicValuedSimplePath) criteriaParameter.getExpressible();
                String parameterName  = expressible.getNavigablePath().getLocalName();
                if (parameterName == null) {
                    throw new ModelException("Can't prepare query for caching as parameter doesn't have a name");
                }
                this.queryParameters.put(parameterName, ((JpaCriteriaParameter<?>) parameter).getValue());
            }
```

We need to figure out how to properly handle this as the above hack won't work in all scenarios.

### Version

quarkus3 branch

### Expected behavior

QueryCacheKey should be able to retrieve the query parameter names.

### Actual behavior

QueryCacheKey obtains names using explicit type casts that won't always work.

### How to Reproduce?

Use quarkus3 branch

**Comment by ahus1** [comment]
**TL;DR:** affects only the the new JPA map store, seems to be manageable. Nothing urgent to contact Hibernate upstream, still it would be nice to get some feedback.

Considering the severity of this: This is about caching queries within a session for the new store. If it proves to difficult, it can be disabled with an performance impact. Even caching some queries but not all queries could be possible. 

I've been analyzing this code and found that this would probably need some additional cases for other types as @sguilhen suggested in the comment, and then it will be fine.

One query that doesn't work is the "users in role" tab for roles. I updated the implementation here: https://github.com/mabartos/keycloak/compare/quarkus3...ahus1:keycloak:is-16332-extract-query-parameter-name?expand=1

Several admin UI endpoints are currently not working as the Admin UI doesn't have a Quarkus 3 branch yet, so I did only minimal explorative testing. Will pause here.

**Comment by ahus1** [comment]
**TL;DR:** updated assessment: query caching with the hack is probably broken. No good solution exists. Will contact Hibernate team.

With the preliminary change done in the initial Quarkus 3 migration, the results are most likely broken as the `queryString` doesn't contain the query, but a dummy value. The second commit in the branch above now contains a workaround to retrieve the HSQL and purify it. Started to prepare an example project based on the Hibernate bug testbed for the Hibernate team to comment.

**Comment by ahus1** [comment]
Created a minimal test case here: https://github.com/hibernate/hibernate-test-case-templates/compare/main...ahus1:hibernate-test-case-templates:is-HHH-xxxx-hql-not-accessible-parameter-name-not-set?expand=1

Discussion started in Hibernate Zulip: https://hibernate.zulipchat.com/#narrow/stream/132096-hibernate-user/topic/H5.20vs.2E.20H6.3A.20Access.20JPQL.20for.20criteria.20builder.20queries

**Comment by ahus1** [comment]
Hibernate team suggests to use the HQL string. After setting an explicit alias, the aliases are cleaned up. Solution doesn't look so ugly any more.
